src/Uno.UI/Helpers/WeakEvents/WeakEventManager.cs

```diff
@@ -6,130 +6,110 @@
-						target.Remove(subscription);
-					}
-					else
-					{
-						subscription.Handler.Invoke(subscriber, null);
-					}
-				}
-			}
+			AddTo(_staticHandlers ??= new(), eventName, handler);
 		}
-
-		public void RemoveEventHandler(Action? handler, [CallerMemberName] string eventName = "")
+		else
 		{
-			if (string.IsNullOrEmpty(eventName))
-				throw new ArgumentNullException(nameof(eventName));
+			AddTo(_handlersPerTarget.GetValue(target, static _ => new()), eventName, handler);
+		}
+	}
 
-			if (handler == null)
-				throw new ArgumentNullException(nameof(handler));
+	public void RemoveEventHandler(Action handler, [CallerMemberName] string eventName = "")
+	{
+		if (string.IsNullOrEmpty(eventName))
+			throw new ArgumentNullException(nameof(eventName));
 
-			RemoveEventHandler(eventName, handler.Target, handler.GetMethodInfo());
-		}
+		if (handler == null)
+			throw new ArgumentNullException(nameof(handler));
 
-		private void AddEventHandler(string eventName, object? handlerTarget, MethodInfo methodInfo)
+		var target = handler.Target;
+		if (target is null)
 		{
-			if (!_eventHandlers.TryGetValue(eventName, out List<Subscription>? targets))
-			{
-				targets = new List<Subscription>();
-				_eventHandlers.Add(eventName, targets);
-			}
-			else
-			{
-				targets.RemoveAll(subscription => subscription.Subscriber is { IsAlive: false });
-			}
-
-			if (handlerTarget == null)
+			if (_staticHandlers is not null)
 			{
-				// This event handler is a static method
-				targets.Add(new Subscription(null, methodInfo));
-				return;
+				RemoveFrom(_staticHandlers, eventName, handler);
 			}
-
-			targets.Add(new Subscription(new WeakReference(handlerTarget), methodInfo));
 		}
-
-		private void RemoveEventHandler(string eventName, object? handlerTarget, MethodInfo methodInfo)
+		else if (_handlersPerTarget.TryGetValue(target, out var handlersPerEvent))
 		{
-			if (!_eventHandlers.TryGetValue(eventName, out List<Subscription>? subscriptions))
-				return;
+			RemoveFrom(handlersPerEvent, eventName, handler);
+		}
+	}
 
-			for (int n = subscriptions.Count - 1; n >= 0; n--)
+	public void HandleEvent(string eventName)
+	{
+		_enumeratingHandlersOfEvent = eventName; // Note: We do not support re-entrancy!
+		try
+		{
+			foreach (var kvp in _handlersPerTarget)
 			{
-				Subscription current = subscriptions[n];
-
-				if (current.Subscriber != null && !current.Subscriber.IsAlive)
+				if (!kvp.Value.TryGetValue(eventName, out var handlers))
 				{
-					// If not alive, remove and continue
-					subscriptions.RemoveAt(n);
+					// Event handlers not found for this target
 					continue;
 				}
 
-				if (current.Subscriber?.Target == handlerTarget && current.Handler == methodInfo)
+				var count = handlers.Count;
+				for (var i = 0; i < count; i++)
 				{
-					// Found the match, we can break
-					subscriptions.RemoveAt(n);
-					break;
+					handlers[i]();
```

Comment:
The handler could end up unsubscribing another handler. This is an existing issue though

Reply:
Yes is has been fixed in https://github.com/unoplatform/uno/pull/16562, that's why pointed me on that class. It's fixed in the add/remove handler: I clone the list at this time (so we will not pay the cost to clone handler's list on each raise unless it's effectively needed!)